File: clawnet-desktop/src/main/store/operation-logger.ts
// src/main/store/operation-logger.ts
//
// 1:1 port of macOS OperationLogger.swift. Workspace-local JSONL persistence
// at <wsRoot>/.clawnet/logs/<UTC-date>.jsonl with full query/find/isUndone API.
//
// Also exports AppAuditLogger (the previous app-level P1E logger) + the
// OperationEntry type for use by agent command handlers.

import { stat, readdir, readFile, writeFile, appendFile, mkdir } from 'node:fs/promises';
import { join } from 'node:path';
import { randomBytes } from 'node:crypto';
import { logsDir, ensureDirectory } from '../utils/workspace-data';
import { LogEntrySchema, type LogEntry, type LogFilter, type LogQueryResult } from '../../shared/domain/operation';

export type { LogEntry, LogFilter, LogQueryResult } from '../../shared/domain/operation';

// ============================================================
// New macOS-parity OperationLogger
// ============================================================

export function generateOperationId(): string {
  return 'op_' + randomBytes(4).toString('hex');
}

function dateString(timestampMs: number): string {
  const d = new Date(timestampMs);
  const y = d.getUTCFullYear();
  const m = String(d.getUTCMonth() + 1).padStart(2, '0');
  const day = String(d.getUTCDate()).padStart(2, '0');
  return `${y}-${m}-${day}`;
}

export class OperationLogger {
  async log(entry: LogEntry, wsRoot: string): Promise<void> {
    const dir = logsDir(wsRoot);
    await ensureDirectory(dir);
    const file = join(dir, dateString(entry.timestamp) + '.jsonl');
    const line = JSON.stringify(entry) + '\n';
    try {
      await stat(file);
      await appendFile(file, line, 'utf-8');
    } catch {
      await writeFile(file, line, 'utf-8');
    }
  }

  async query(filter: LogFilter, wsRoot: string): Promise<LogQueryResult> {
    const dir = logsDir(wsRoot);
    let files: string[];
    try { files = await readdir(dir); } catch { return { entries: [], total: 0, hasMore: false }; }
    // Default: no time restriction (0 = epoch, MAX_SAFE_INTEGER = far future).
    // Callers that want today-only behaviour should pass explicit since/until.
    const since = filter.since ?? 0;
    const until = filter.until ?? Number.MAX_SAFE_INTEGER;
    const sinceDate = dateString(since);
    const untilDate = dateString(until);

    const all: LogEntry[] = [];
    const sortedFiles = files.filter((f) => f.endsWith('.jsonl')).sort();
    for (const f of sortedFiles) {
      const dateStr = f.slice(0, -6); // remove ".jsonl"
      if (dateStr < sinceDate || dateStr > untilDate) continue;
      let content: string;
      try { content = await readFile(join(dir, f), 'utf-8'); } catch { continue; }
      for (const line of content.split('\n')) {
        if (!line) continue;
        let parsedJson: unknown;
        try { parsedJson = JSON.parse(line); } catch { continue; }
        const parsed = LogEntrySchema.safeParse(parsedJson);
        if (!parsed.success) continue;
        const e = parsed.data;
        if (e.timestamp < since || e.timestamp > until) continue;
        if (filter.sessionId !== undefined && e.sessionId !== filter.sessionId) continue;
        if (filter.command !== undefined && e.command !== filter.command) continue;
        all.push(e);
      }
    }

    all.sort((a, b) => b.timestamp - a.timestamp);
    const limit = filter.limit;
    const offset = Math.min(filter.offset, all.length);
    const end = Math.min(offset + limit, all.length);
    return { entries: all.slice(offset, end), total: all.length, hasMore: end < all.length };
  }

  async findEntry(operationId: string, wsRoot: string): Promise<LogEntry | null> {
    const dir = logsDir(wsRoot);
    let files: string[];
    try { files = await readdir(dir); } catch { return null; }
    const sorted = files.filter((f) => f.endsWith('.jsonl')).sort().reverse();
    for (const f of sorted) {
      let content: string;
      try { content = await readFile(join(dir, f), 'utf-8'); } catch { continue; }
      for (const line of content.split('\n')) {
        if (!line) continue;
        let parsedJson: unknown;
        try { parsedJson = JSON.parse(line); } catch { continue; }
        const parsed = LogEntrySchema.safeParse(parsedJson);
        if (parsed.success && parsed.data.id === operationId) return parsed.data;
      }
    }
    return null;
  }

  async isUndone(operationId: string, wsRoot: string): Promise<boolean> {
    const dir = logsDir(wsRoot);
    let files: string[];
    try { files = await readdir(dir); } catch { return false; }
    const sorted = files.filter((f) => f.endsWith('.jsonl')).sort().reverse();
    for (const f of sorted) {
      let content: string;
      try { content = await readFile(join(dir, f), 'utf-8'); } catch { continue; }
      for (const line of content.split('\n')) {
        if (!line) continue;
        let parsedJson: unknown;
        try { parsedJson = JSON.parse(line); } catch { continue; }
        const parsed = LogEntrySchema.safeParse(parsedJson);
        if (parsed.success && parsed.data.type === 'undo' && parsed.data.undoTargetId === operationId) {
          return true;
        }
      }
    }
    return false;
  }
}

// ============================================================
// P1E compatibility: app-level audit logger for consent-UI flow
// ============================================================

export interface OperationEntry {
  kind: string;
  agentId?: string;
  path?: string;
  op?: string;
  decision?: string;
  reason?: string;
  details?: Record<string, unknown>;
}

interface StoredEntry extends OperationEntry {
  timestamp: string;
}

export class AppAuditLogger {
  constructor(private readonly logsDir: string) {}
  async record(entry: OperationEntry): Promise<void> {
    const now = new Date();
    const stored: StoredEntry = { ...entry, timestamp: now.toISOString() };
    const y = now.getUTCFullYear();
    const m = String(now.getUTCMonth() + 1).padStart(2, '0');
    const d = String(now.getUTCDate()).padStart(2, '0');
    const filename = `ops-${y}-${m}-${d}.jsonl`;
    const filePath = join(this.logsDir, filename);
    await mkdir(this.logsDir, { recursive: true });
    await appendFile(filePath, JSON.stringify(stored) + '\n', { encoding: 'utf-8' });
  }
}

File: clawnet-macosapp/ClawNet/Services/OperationLogger.swift
import Foundation
import OSLog

/// JSONL-based operation logger for file change tracking.
/// Logs are stored in `.clawnet/logs/` with one file per day, append-only.
@MainActor
final class OperationLogger {
    static let shared = OperationLogger()

    private let logger = Logger(subsystem: "ai.clawnet.macos", category: "operation-logger")

    /// Commands that must be logged (all mutating file operations).
    static let loggableCommands: Set<String> = [
        "file.move", "file.rename", "file.copy", "file.write", "file.trash", "file.mkdir",
    ]

    // MARK: - Data Models

    struct LogEntry: Codable {
        let id: String
        let timestamp: Int64
        let sessionId: String?
        let command: String
        let params: [String: JSONValue]
        let result: String              // "success" or "error"
        let errorMessage: String?
        let reversible: Bool
        let reverseAction: ReverseAction?

        /// Entry type: nil = normal operation, "undo" = undo record, "rollback" = rollback record
        let type: String?
        /// For type=="undo": the operation ID that was undone
        let undoTargetId: String?
    }

    struct ReverseAction: Codable {
        let command: String
        let params: [String: JSONValue]
    }

    struct LogFilter {
        var sessionId: String?
        var command: String?
        var since: Int64?
        var until: Int64?
        var limit: Int = 50
        var offset: Int = 0
    }

    struct LogQueryResult {
        let entries: [LogEntry]
        let total: Int
        let hasMore: Bool
    }

    // MARK: - ID Generation

    static func generateId() -> String {
        let hex = (0..<4).map { _ in String(format: "%02x", UInt8.random(in: 0...255)) }.joined()
        return "op_\(hex)"
    }

    // MARK: - Write

    func log(_ entry: LogEntry, wsRoot: URL) {
        do {
            let logsDir = try ClawNetDataManager.ensureDirectory(ClawNetDataManager.logsDir(wsRoot: wsRoot))
            let fileName = Self.dateString(from: entry.timestamp) + ".jsonl"
            let fileURL = logsDir.appendingPathComponent(fileName)

            let data = try JSONEncoder().encode(entry)
            guard var line = String(data: data, encoding: .utf8) else { return }
            line += "\n"

            if FileManager.default.fileExists(atPath: fileURL.path) {
                let handle = try FileHandle(forWritingTo: fileURL)
                defer { try? handle.close() }
                handle.seekToEndOfFile()
                handle.write(Data(line.utf8))
            } else {
                try Data(line.utf8).write(to: fileURL, options: [.atomic])
            }
        } catch {
            logger.error("Failed to write log: \(error.localizedDescription, privacy: .public)")
        }
    }

    // MARK: - Query

    func query(filter: LogFilter, wsRoot: URL) -> LogQueryResult {
        let logsDir = ClawNetDataManager.logsDir(wsRoot: wsRoot)
        let fm = FileManager.default

        guard fm.fileExists(atPath: logsDir.path) else {
            return LogQueryResult(entries: [], total: 0, hasMore: false)
        }

        // Determine date range for files to read
        let now = Int64(Date().timeIntervalSince1970 * 1000)
        let sinceMs = filter.since ?? Self.startOfDayMs()
        let untilMs = filter.until ?? now

        let sinceDate = Self.dateString(from: sinceMs)
        let untilDate = Self.dateString(from: untilMs)

        // Collect all matching log files
        var allEntries: [LogEntry] = []
        if let files = try? fm.contentsOfDirectory(atPath: logsDir.path) {
            let sortedFiles = files.filter { $0.hasSuffix(".jsonl") }.sorted()
            for file in sortedFiles {
                let dateStr = String(file.dropLast(6)) // remove .jsonl
                if dateStr < sinceDate || dateStr > untilDate { continue }

                let fileURL = logsDir.appendingPathComponent(file)
                guard let data = try? String(contentsOf: fileURL, encoding: .utf8) else { continue }

                let decoder = JSONDecoder()
                for line in data.components(separatedBy: "\n") where !line.isEmpty {
                    guard let lineData = line.data(using: .utf8),
                          let entry = try? decoder.decode(LogEntry.self, from: lineData) else { continue }

                    // Apply filters
                    if entry.timestamp < sinceMs || entry.timestamp > untilMs { continue }
                    if let sid = filter.sessionId, entry.sessionId != sid { continue }
                    if let cmd = filter.command, entry.command != cmd { continue }

                    allEntries.append(entry)
                }
            }
        }

        // Sort by timestamp descending (most recent first)
        allEntries.sort { $0.timestamp > $1.timestamp }

        let total = allEntries.count
        let offset = min(filter.offset, total)
        let end = min(offset + filter.limit, total)
        let paged = Array(allEntries[offset..<end])

        return LogQueryResult(entries: paged, total: total, hasMore: end < total)
    }

    // MARK: - Lookup

    /// Find a specific log entry by ID across all log files in the workspace.
    func findEntry(operationId: String, wsRoot: URL) -> LogEntry? {
        let logsDir = ClawNetDataManager.logsDir(wsRoot: wsRoot)
        guard let files = try? FileManager.default.contentsOfDirectory(atPath: logsDir.path) else { return nil }

        let decoder = JSONDecoder()
        // Search in reverse chronological order (most recent files first)
        for file in files.filter({ $0.hasSuffix(".jsonl") }).sorted().reversed() {
            let fileURL = logsDir.appendingPathComponent(file)
            guard let data = try? String(contentsOf: fileURL, encoding: .utf8) else { continue }
            for line in data.components(separatedBy: "\n") where !line.isEmpty {
                guard let lineData = line.data(using: .utf8),
                      let entry = try? decoder.decode(LogEntry.self, from: lineData) else { continue }
                if entry.id == operationId { return entry }
            }
        }
        return nil
    }

    /// Check if an operation has been undone (exists an undo entry targeting it).
    func isUndone(operationId: String, wsRoot: URL) -> Bool {
        let logsDir = ClawNetDataManager.logsDir(wsRoot: wsRoot)
        guard let files = try? FileManager.default.contentsOfDirectory(atPath: logsDir.path) else { return false }

        let decoder = JSONDecoder()
        for file in files.filter({ $0.hasSuffix(".jsonl") }).sorted().reversed() {
            let fileURL = logsDir.appendingPathComponent(file)
            guard let data = try? String(contentsOf: fileURL, encoding: .utf8) else { continue }
            for line in data.components(separatedBy: "\n") where !line.isEmpty {
                guard let lineData = line.data(using: .utf8),
                      let entry = try? decoder.decode(LogEntry.self, from: lineData) else { continue }
                if entry.type == "undo" && entry.undoTargetId == operationId { return true }
            }
        }
        return false
    }

    // MARK: - Helpers

    private static let dateFormatter: DateFormatter = {
        let f = DateFormatter()
        f.dateFormat = "yyyy-MM-dd"
        f.timeZone = TimeZone.current
        return f
    }()

    static func dateString(from timestampMs: Int64) -> String {
        let date = Date(timeIntervalSince1970: Double(timestampMs) / 1000.0)
        return dateFormatter.string(from: date)
    }

    static func startOfDayMs() -> Int64 {
        let cal = Calendar.current
        let start = cal.startOfDay(for: Date())
        return Int64(start.timeIntervalSince1970 * 1000)
    }
}

// MARK: - JSON Value (type-erased Codable)

/// A simple type-erased JSON value for storing arbitrary params in log entries.
enum JSONValue: Codable, Equatable {
    case string(String)
    case int(Int64)
    case double(Double)
    case bool(Bool)
    case null

    init(from decoder: Decoder) throws {
        let container = try decoder.singleValueContainer()
        if let v = try? container.decode(Bool.self) { self = .bool(v) }
        else if let v = try? container.decode(Int64.self) { self = .int(v) }
        else if let v = try? container.decode(Double.self) { self = .double(v) }
        else if let v = try? container.decode(String.self) { self = .string(v) }
        else if container.decodeNil() { self = .null }
        else { self = .null }
    }

    func encode(to encoder: Encoder) throws {
        var container = encoder.singleValueContainer()
        switch self {
        case .string(let v): try container.encode(v)
        case .int(let v): try container.encode(v)
        case .double(let v): try container.encode(v)
        case .bool(let v): try container.encode(v)
        case .null: try container.encodeNil()
        }
    }

    var stringValue: String? {
        if case .string(let v) = self { return v }
        return nil
    }

    var boolValue: Bool? {
        if case .bool(let v) = self { return v }
        return nil
    }
}

/// Convert [String: Any] params to [String: JSONValue] for logging.
func paramsToJSONValues(_ params: [String: Any]) -> [String: JSONValue] {
    var result: [String: JSONValue] = [:]
    for (key, value) in params {
        switch value {
        case let v as String: result[key] = .string(v)
        case let v as Bool: result[key] = .bool(v)
        case let v as Int: result[key] = .int(Int64(v))
        case let v as Int64: result[key] = .int(v)
        case let v as Double: result[key] = .double(v)
        default: result[key] = .null
        }
    }
    return result
}
